request/quic_transport.go

```diff
@@ -29,28 +37,52 @@ func (m *QuicTransport) RoundTrip(req *http.Request) (*http.Response, error) {
 	req.ProtoMajor = 3
 	req.ProtoMinor = 0
 	req.Proto = "HTTP/3"
-	inner := m.inner.Load()
-	resp, err := inner.RoundTrip(req)
-	if err != nil &&
-		(strings.Contains(err.Error(), "Application error 0x100") ||
-			strings.Contains(err.Error(), "no recent network activity") ||
-			strings.Contains(err.Error(), "Timeout exceeded while awaiting headers")) {
-		// connection closed, need to reconnect
-		if inner != nil {
-			inner.Close()
-		}
-		inner := m.createFn()
-		m.inner.Store(inner)
-		resp, err = inner.RoundTrip(req)
-	}
+	resp, err := m.executeRequest(req)
 	return resp, err
 }
 
 func (m *QuicTransport) NotifyConnect(events.DataConnect) error {
-	inner := m.inner.Load()
-	if inner != nil {
-		inner.Close()
-	}
-	m.inner.Store(m.createFn())
+	m.recreateRoundTrip(true)
 	return nil
 }
+
+func (m *QuicTransport) executeRequest(req *http.Request) (*http.Response, error) {
+	m.mu.RLock()
+	response, err := m.inner.RoundTrip(req)
+
+	// check the errors if inner needs to be recreated
+	recreate := shouldRecreate(err)
+
+	// mark that inner needs to be recreated while holding the read lock to be sure shouldRecreate is not replaced before recreation
+	m.shouldRecreate.Store(recreate)
+	m.mu.RUnlock()
+
+	if recreate {
+		log.Println(internal.InfoPrefix, "recreate RoundTripper on error", err)
+		// recreate the inner and retry one more time to execute the same request
+		m.recreateRoundTrip(false)
+		m.mu.RLock()
+		response, err = m.inner.RoundTrip(req)
+		m.mu.RUnlock()
+	}
+
+	return response, err
+}
+
+func shouldRecreate(err error) bool {
```

Comment:
I recommend to rename this fn as it is the same as variable and creates unnecessary code reading confusion, maybe it could be renamed to `checkShouldRecreate`